### PR TITLE
Kibana Deployment

### LINKED ISSUE #2790: Kibana Deployment
**Description:**
- Update manifests to support deploying Kibana via  docker
   - buildpacks do not support cfos4

**Acceptance Criteria:** 

- [ ]   Kibana deploys to dev environment
- [ ]  Routes between Kibana and Elastic are configured
- [ ]  Kibana can connect to Elastic
- [ ]  User can navigate to Kibana from frontend
- [ ] Testing Checklist has been run and all tests pass
- [ ] README is updated, if necessary

**Tasks:**
_Create a list of granular, specific work items that must be completed to deliver the desired outcomes of this issue_ 

- [ ] Update backend app manifest to deploy Kibana as a docker container
  - [ ] Parse hostname, username, and password from VCAP env vars and feed to container
  - [ ] xpack should be disabled
- [ ] Verify Kibana and Elastic are connected and communicating securely 
- [ ] Create necessary routes for the frontend and backend to talk to Kibana and redirect to Kibana
- [ ] Verify only HHS AMS authenticated admin users can access Kibana
- [ ] Run Testing Checklist and confirm all tests pass

**Notes:**
_Add additional useful information, such as related issues and functionality that isn't covered by this specific issue, and other considerations that will be helpful for anyone reading this_ 

- _Note 1_
- _Note 2_
- _Note 3_

**Supporting Documentation:**
_Please include any relevant log snippets/files/screen shots_

- _Doc 1_
- _Doc 2_

**Open Questions:**
_Please include any questions or decisions that must be made before beginning work or to confidently call this issue complete_ 

- _Open Question 1_
- _Open Question 2_

### PR DESCRIPTION
**Pull request by elipe17**
## Summary of Changes
- Added manifests to deploy Kibana and the Elastic proxy
- Updated Nginx to act as an auth endpoint for Kibana
- Removed elastic_setup container because it is now irrelevant
- Updated deployment configs
Pull request closes #2790 

## How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose up
cd tdrs-backend && docker-compose up
```

1. Open http://localhost:3000/ and sign in.
2. Navigate to /kibana and verify you get an error from nginx
3. Add BYPASS_KIBANA_AUTH=yes to your .env
4. Navigate to /kibana and verify you can see Kibana and it feels native to frontend
5. Navigate to https://tdp-frontend-raft.app.cloud.gov and perform the previous steps
  - To see Kibana in the deployed environment follow the steps below to give yourself access. Make sure you have already requested access by way of the frontend. If you already have access, the steps below still apply. If you don't have the cf cli tool, let me know and I can perform the steps for you after you have requested access from the frontend.
  1. Login to cloud foundry with the CLI: `cf login --sso`
  2. Ssh into backend: `cf ssh tdp-backend-raft`
  3. Enable python: `/tmp/lifecycle/shell`
  4. Start Django shell: `python manage.py shell_plus`
  5. Get your user: `user = User.objects.all().filter(email="YOUR_LOGIN_EMAIL_HERE")[0]`
  6. Approve your user: `user.account_approval_status = "Approved"`
  7. Get correct group: `group = Group.objects.get(name="OFA System Admin")`
  8. Add yourself to group: `group.user_set.add(user)`
  9. Give user a fake hhs_id: `user.hhs_id = "111111111111"`
  10. Save: `user.save()`
  11. Verify you can see the frontend now and the Kibana tab.

## Deliverables
_More details on how deliverables herein are assessed included [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverables)._

### [Deliverable 1: Accepted Features](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-1-Accepted-Features)

Checklist of ACs:
- [x]   Kibana deploys to dev environment
- [x]  Routes between Kibana and Elastic are configured
- [x]  Kibana can connect to Elastic
- [x]  User can navigate to Kibana from frontend
- [x] Testing Checklist has been run and all tests pass
- [x] README is updated, if necessary
+ [x] **`lfrohlich`** and/or **`adpennington`**  confirmed that ACs are met.

### [Deliverable 2: Tested Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-2-Tested-Code)

+ Are all areas of code introduced in this PR meaningfully tested?
  + [x] If this PR introduces backend code changes, are they meaningfully tested?
  + [x] If this PR introduces frontend code changes, are they meaningfully tested?
+ Are code coverage minimums met?
  + [x] Frontend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)
  + [x] Backend coverage: [_insert coverage %_] (see `CodeCov Report` comment in PR)

### [Deliverable 3: Properly Styled Code](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-3-Properly-Styled-Code)

+ [x] Are backend code style checks passing on CircleCI?
+ [x] Are frontend code style checks passing on CircleCI?
+ [x] Are code maintainability principles being followed?

### [Deliverable 4: Accessible](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-4-Accessibility)

+ [ ] Does this PR complete the epic? 
+ [ ] Are links included to any other gov-approved PRs associated with epic?
+ [ ] Does PR include documentation for Raft's a11y review? 
+ [ ] Did automated and manual testing with `iamjolly` and `ttran-hub` using Accessibility Insights reveal any errors introduced in this PR?

### [Deliverable 5: Deployed](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-5-Deployed)

+ [x] Was the code successfully deployed via automated CircleCI process to development on Cloud.gov?

### [Deliverable 6: Documented](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-6-Code-documentation)

+ [x] Does this PR provide background for why coding decisions were made?
+ [ ] If this PR introduces backend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces frontend code, is that code easy to understand and sufficiently documented, both inline and overall?
+ [ ] If this PR introduces dependencies, are their licenses documented?
+ [x] Can reviewer explain and take ownership of these elements presented in this code review?

### [Deliverable 7: Secure](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-7-Secure)

+ [x] Does the OWASP Scan pass on CircleCI?
+ [ ] Do manual code review and manual testing detect any new security issues?
+ [ ] If new issues detected, is investigation and/or remediation plan documented? 

### [Deliverable 8: User Research](https://github.com/raft-tech/TANF-app/blob/develop/docs/How-We-Work/our-priorities-values-expectations.md#Deliverable-8-User-Research)

Research product(s) clearly articulate(s):
+ [ ] the purpose of the research
+ [ ] methods used to conduct the research 
+ [ ] who participated in the research
+ [ ] what was tested and how
+ [ ] impact of research on TDP
+ [ ] (_if applicable_) final design mockups produced for TDP development

### CODE REVIEW

File: tdrs-frontend/nginx/cloud.gov/locations.conf
Comment (raftmsohani) [inline tdrs-frontend/nginx/cloud.gov/locations.conf]: I don't think we need buffer for auth, also not the " proxy_temp_file_write_size"

**Review by raftmsohani — APPROVED** [review]
LGTM! great work 👍

**Review by andrew-jameson — APPROVED** [review]

**Review by jtimpe — APPROVED** [review]
Excellent work! 🚀

File: tdrs-frontend/nginx/cloud.gov/buildpack.nginx.conf
Comment (ADPennington) [inline tdrs-frontend/nginx/cloud.gov/buildpack.nginx.conf:66]: fyi:  this unsafe-inline is showing up in security scans and if we need this setting, we should document why. the only documentation im aware of on this lives [here](https://github.com/raft-tech/TANF-app/blob/develop/docs/Technical-Documentation/Architecture-Decision-Record/016-zap-waiver.md), and im not sure if the justification is still accurate.

File: tdrs-backend/tdpservice/users/api/authorization_check.py
Comment (ADPennington) [inline tdrs-backend/tdpservice/users/api/authorization_check.py:69]: noticing this line of code trigger 8-10x in the logs after clicking kibana endpoint.

File: tdrs-frontend/nginx/cloud.gov/buildpack.nginx.conf
Comment (ADPennington) [reply tdrs-frontend/nginx/cloud.gov/buildpack.nginx.conf:64]: related to the above: it stops after the endpoint page loads.

**Review by ADPennington — COMMENTED** [review]
@elipe17 this looks good. Some summary-level notes/comments below. My testing notes are captured in more detail here: https://hackmd.io/P4sQkTRORLGjiodfgRwMtA?both#Kibana-Deployment

* If this will introduce new apps and resources (e.g. services) into the prod environment, we need the system diagrams and boundary diagram updated. This can be done in a follow-up ticket when this feature is more refined. 
* The Kibana tab is visible (the endpoint is not accessible) for `Developer` and `OFA System Admin` roles when signing in via login.gov. 
* We will still need to do some additional testing in staging to better understand what is causing the kibana routing issue.  
* I need more information on why we're using `unsafe-inline` and `unsafe-eval` settings since this often comes up in the security scans and we need to explain why to security team. 
* conflicts need to be resolved herein before merge.

File: tdrs-backend/tdpservice/users/api/authorization_check.py
Comment (elipe17) [reply tdrs-backend/tdpservice/users/api/authorization_check.py:69]: This log line will be hit every time the frontend makes a request to Kibana. This is because we need to pass through the authentication view every time prior to proxying the request to Kibana to ensure the user is still passing the traditional Auth Check and their group hasnt changed. Debug is the highest level for the logger so it wont show up in prod.

File: tdrs-frontend/nginx/cloud.gov/buildpack.nginx.conf
Comment (elipe17) [reply tdrs-frontend/nginx/cloud.gov/buildpack.nginx.conf:66]: Unfortunately this is a requirement with Kibana and it likely will be for a long time. Kibana uses libraries that require the `unsafe-*`policies. The version of Kibana we are using was released in [October, 2019](https://www.elastic.co/blog/kibana-7-4-0-released) and [THIS TOPIC](https://discuss.elastic.co/t/does-kibana-need-the-autorisation-of-unsafe-inline-or-unsafe-eval-to-work-properly/234390#:~:text=May%202020-,Hey,-%40Otmane_Faouzi) states that as of May, 2020 the the requirement of these policies still exists. [THIS ISSUE](https://github.com/elastic/kibana/issues/132595l) captures the work that would allow the removal of those policies. However, it was opened in 2022 and is still open. 

Would you like me to update the documentation you linked, or should I create a new doc capturing this information?

**Comment by elipe17** [comment]
@ADPennington I updated based on your feedback/responded to your questions. I updated the frontend to only display the Kibana tab if you are authenticated via AMS along with the existing controls. I also added a switch for local dev to display it since we cannot authenticate that way.

File: tdrs-frontend/nginx/cloud.gov/buildpack.nginx.conf
Comment (ADPennington) [inline tdrs-frontend/nginx/cloud.gov/buildpack.nginx.conf:37]: @elipe17 just noting for your reference that after adding missing env vars +  re-staging backend, i still had to restage frontend to authenticate. im not sure if this resolves the issue described in #2839

File: tdrs-frontend/nginx/cloud.gov/buildpack.nginx.conf
Comment (ADPennington) [reply tdrs-frontend/nginx/cloud.gov/buildpack.nginx.conf:66]: @elipe17 I think this is worth addressing as part of #2843.

File: tdrs-frontend/src/selectors/auth.js
Comment (ADPennington) [inline tdrs-frontend/src/selectors/auth.js:65]: not sure its worth noting as an inline comment here that the developer role + acf email yields the kibana tabs but endpoint is still inaccessible. this might only be relevant to dev team and qasp reviewers since developer role doesnt exist in prod. @elipe17

Comment (ADPennington) [reply tdrs-frontend/src/selectors/auth.js:65]: 2805p3

Comment (elipe17) [reply tdrs-frontend/src/selectors/auth.js:65]: Good catch, I will remove developer!

**Review by ADPennington — APPROVED** [review]
excellent work @elipe17 thanks for all the testing support! 🚀 

- added system diagram updates as a task to #1441 
- #2839 will capture work for troubleshooting routing issue
- #2843 will capture updates to security tool ADR re: our use of unsafe-* policies

**Pull request merged**
